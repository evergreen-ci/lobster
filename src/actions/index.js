// @flow strict

import type { Action as LogviewerAction } from './logviewer';
import type { Log } from '../models';

export const LOGKEEPER_LOAD_DATA = 'logkeeper:load-data';
export const LOBSTER_LOAD_DATA = 'lobster:load-data';
export const PROCESS_RESPONSE = 'process-response';
export const LOAD_CACHED_DATA = 'load-cached-data';
export const EVERGREEN_LOAD_DATA = 'evergreen:load-data';
export const SETUP_CACHE = 'setup-cache';
export const WIPE_CACHE = 'wipe-cache';

export type LogkeeperLoadData = {|
  +type: 'logkeeper:load-data',
  +payload: {|
    +build: string,
    +test: ?string
  |}
|}

export type LobsterLoadData = {|
  +type: 'lobster:load-data',
  +payload: {|
    +url: string,
    +server: string
  |}
|}

const logTypes_ = {
  'resmoke': '',
  'raw': ''
};

export const logTypes = (): string[] => Object.keys(logTypes_);

export type LogType = $Keys<typeof logTypes_>

export type ProcessResponse = {|
  +type: 'process-response',
  +payload: {|
    +type: LogType,
    +data: string,
    +isDone: boolean
  |},
  +error: boolean
|}

export function logkeeperLoadData(build: string, test: ?string): LogkeeperLoadData {
  return {
    type: LOGKEEPER_LOAD_DATA,
    payload: {
      build: build,
      test: test
    }
  };
}

// Load data from the lobster server
export function lobsterLoadData(server: string, url: string): LobsterLoadData {
  return {
    type: LOBSTER_LOAD_DATA,
    payload: {
      url: url,
      server: server
    }
  };
}

export function processData(data: string, type: LogType, isDone?: boolean): ProcessResponse {
  return {
    type: PROCESS_RESPONSE,
    payload: {
      type: type,
      data: data,
      isDone: isDone || false
    },
    error: false
  };
}

export function processLocalData(data: string, type: LogType): ProcessResponse {
  return {
    type: PROCESS_RESPONSE,
    payload: {
      type: type,
      data: data,
      isDone: true
    },
    error: false
  };
}

export function processDataError(data: string): ProcessResponse {
  return {
    type: PROCESS_RESPONSE,
    payload: {
      type: 'resmoke',
      data: data,
      isDone: true
    },
    error: true
  };
}

export type LoadCachedData = {|
  type: 'load-cached-data',
  payload: {|
    log: Log
  |}
|}

export function loadCachedData(data: Log): LoadCachedData {
  return {
    type: LOAD_CACHED_DATA,
    payload: {
      log: data
    }
  };
}

export type CacheStatus = 'ok' | 'error' | 'never' | 'later' | 'unsupported' | null;

export type SetupCache = {|
  +type: 'setup-cache',
  +payload: {
    +status: CacheStatus,
    +size: number
  },
  +error: boolean
|}

export function setCache(status: CacheStatus, size: number): SetupCache {
  return {
    type: SETUP_CACHE,
    payload: {
      status: status,
      size: size
    },
    error: false
  };
}

const evergreenTaskLogTypes: { [string]: string } = {
  'all': 'ALL',
  'task': 'T',
  'agent': 'A',
  'system': 'S'
  // 'event': 'E' // Not actually supported by the api
};

export type EvergreenTaskLogType = $Keys<typeof evergreenTaskLogTypes>;

export function stringToInteralEvergreenTaskLogType(a: string): ?string {
  return evergreenTaskLogTypes[a];
}

export function stringToEvergreenTaskLogType(a: string): ?EvergreenTaskLogType {
  if (!evergreenTaskLogTypes[a]) {
    return null;
  }

  return a;
}

export type EvergreenTaskLog = {|
  type: 'task',
  id: string,
  execution: number,
  log: EvergreenTaskLogType
|};

export type EvergreenTestLog = {|
  type: 'test',
  id: string,
|};

export type EvergreenLoadData = {|
  +type: 'evergreen:load-data',
  +payload: EvergreenTaskLog | EvergreenTestLog
|}

export function evergreenLoadTaskLog(id: string, execution: number, log: EvergreenTaskLogType): EvergreenLoadData {
  return {
    type: 'evergreen:load-data',
    payload: {
      type: 'task',
      id: id,
      execution: execution,
      log: log
    }
  };
}

export function evergreenLoadTestLog(id: string): EvergreenLoadData {
  return {
    type: 'evergreen:load-data',
    payload: {
      type: 'test',
      id: id
    }
  };
}

export type WipeCache = {|
  type: 'wipe-cache',
  payload: {|
    file: ?string
  |}
|}

export function wipeCache(): WipeCache {
  return {
    type: WIPE_CACHE,
    payload: {
      file: null
    }
  };
}

export function fromFileFromCache(f: string): WipeCache {
  return {
    type: WIPE_CACHE,
    payload: {
      file: f
    }
  };
}
export type Action = ProcessResponse
  | LoadCachedData
  | WipeCache
  | LogkeeperLoadData
  | LobsterLoadData
  | EvergreenLoadData
  | LogviewerAction
