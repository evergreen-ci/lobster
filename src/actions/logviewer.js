// @flow strict

export const LOGVIEWER_CHANGE_SETTING = 'logviewer:change-setting';
export const LOGVIEWER_CHANGE_FILTER = 'logviewer:change-filter';
export const LOGVIEWER_LOAD_FILTERS = 'logviewer:load-filters';
export const LOGVIEWER_CHANGE_HIGHLIGHT = 'logviewer:change-highlight';
export const LOGVIEWER_LOAD_HIGHLIGHTS = 'logviewer:load-highlights';
export const LOGVIEWER_CHANGE_BOOKMARK = 'logviewer:change-bookmark';
export const LOGVIEWER_LOAD_BOOKMARKS = 'logviewer:load-bookmarks';
export const LOGVIEWER_ENSURE_BOOKMARK = 'logviewer:ensure-bookmark';
export const LOGVIEWER_CHANGE_FINDIDX = 'logviewer:change-findidx';
export const LOGVIEWER_CHANGE_SEARCH = 'logviewer:change-search';

export type Filter = {|
  text: string,
  on: boolean,
  inverse: boolean
|}

export type Highlight = {|
  text: string,
  on: boolean,
  line: boolean
|}

export type Bookmark = {|
  lineNumber: number,
|}

export type ChangeSetting = {|
  +type: 'logviewer:change-setting',
  +payload: {|
    +setting: string,
    +value: string
  |}
|}

export type ChangeFilter = {|
  +type: 'logviewer:change-filter',
  +payload: {|
    +field: string,
    +text: string
  |}
|}

export type ChangeHighlight = {|
  +type: 'logviewer:change-highlight',
  +payload: {|
    +field: string,
    +text: string
  |}
|}

export type ChangeBookmark = {|
  +type: 'logviewer:change-bookmark',
  +payload: {|
    +lineNumArray: number[]
  |}
|}

export type EnsureBookmark = {|
  +type: 'logviewer:ensure-bookmark',
  +payload: {|
    +lineNum: number
  |}
|}

export type LoadBookmarks = {|
  +type: 'logviewer:load-bookmarks',
  +payload: {|
    +bookmarksArr: Bookmark[]
  |}
|}

export type ChangeFindIdx = {|
  +type: 'logviewer:change-findidx',
  +payload: {|
    +index: number
  |}
|}

export type ChangeSearch = {|
  +type: 'logviewer:change-search',
  +payload: {|
    +text: string
  |}
|}

export type LoadHighlights = {|
  +type: 'logviewer:load-highlights',
  +payload: {|
    +initialHighlights: Highlight[]
  |}
|}

export type LoadFilters = {|
  +type: 'logviewer:load-filters',
  +payload: {|
    +initialFilters: Filter[]
  |}
|}

export type Action = ChangeSetting
  | ChangeFilter
  | ChangeHighlight
  | ChangeBookmark
  | EnsureBookmark
  | LoadBookmarks
  | ChangeFindIdx
  | ChangeSearch
  | LoadHighlights
  | LoadFilters

function toggleSetting(setting: string): ChangeSetting {
  return {
    type: LOGVIEWER_CHANGE_SETTING,
    payload: {
      setting: setting,
      value: 'toggle'
    }
  };
}

export function toggleLineWrap(): ChangeSetting {
  return toggleSetting('line-wrap');
}

export function toggleCaseSensitivity(): ChangeSetting {
  return toggleSetting('case-sensitive');
}

export function toggleFilterIntersection(): ChangeSetting {
  return toggleSetting('filter-intersection');
}

export function loadInitialFilters(initialFilters: Filter[]): loadInitialFilters {
  return {
    type: LOGVIEWER_LOAD_FILTERS,
    payload: {
      initialFilters: initialFilters
    }
  };
}

function changeFilter(field: string, text: string): ChangeFilter {
  return {
    type: LOGVIEWER_CHANGE_FILTER,
    payload: {
      field: field,
      text: text
    }
  };
}

export function addFilter(text: string): ChangeFilter {
  return changeFilter('add', text);
}

export function toggleFilterInverse(text: string): ChangeFilter {
  return changeFilter('inverse', text);
}

export function toggleFilter(text: string): ChangeFilter {
  return changeFilter('on', text);
}

export function removeFilter(text: string): ChangeFilter {
  return changeFilter('remove', text);
}

export function loadInitialHighlights(initialHighlights: Highlight[]): loadInitialHighlights {
  return {
    type: LOGVIEWER_LOAD_HIGHLIGHTS,
    payload: {
      initialHighlights: initialHighlights
    }
  };
}

function changeHighlight(field: string, text: string): ChangeHighlight {
  return {
    type: LOGVIEWER_CHANGE_HIGHLIGHT,
    payload: {
      field: field,
      text: text
    }
  };
}

export function addHighlight(text: string): ChangeHighlight {
  return changeHighlight('add', text);
}

export function toggleHighlightLine(text: string): ChangeHighlight {
  return changeHighlight('line', text);
}

export function toggleHighlight(text: string): ChangeHighlight {
  return changeHighlight('on', text);
}

export function removeHighlight(text: string): ChangeHighlight {
  return changeHighlight('remove', text);
}

export function toggleBookmark(lineNumArray: number[]): ChangeBookmark {
  return {
    type: LOGVIEWER_CHANGE_BOOKMARK,
    payload: {
      lineNumArray: lineNumArray
    }
  };
}

export function ensureBookmark(lineNum: number): EnsureBookmark {
  return {
    type: LOGVIEWER_ENSURE_BOOKMARK,
    payload: {
      lineNum: lineNum
    }
  };
}

export function loadBookmarks(bookmarksArr: Bookmark[]): LoadBookmarks {
  return {
    type: LOGVIEWER_LOAD_BOOKMARKS,
    payload: {
      bookmarksArr: bookmarksArr
    }
  };
}

export function changeFindIdx(index: number): ChangeFindIdx {
  return {
    type: LOGVIEWER_CHANGE_FINDIDX,
    payload: {
      index: index
    }
  };
}

export function changeSearch(text: string): ChangeSearch {
  return {
    type: LOGVIEWER_CHANGE_SEARCH,
    payload: {
      text: text
    }
  };
}
