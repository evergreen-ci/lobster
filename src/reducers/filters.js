// @flow strict

import { LOGVIEWER_CHANGE_FILTER, LOGVIEWER_LOAD_FILTERS } from '../actions/logviewer';
import type { Action } from '../actions';
import type { Filter } from '../actions/logviewer';

const initialState: Filter[] = [];

export default function(state: Filter[] = initialState, action: Action): Filter[] {
  if (action.type === LOGVIEWER_LOAD_FILTERS) {
    return action.payload.initialFilters;
  }

  if (action.type !== LOGVIEWER_CHANGE_FILTER) {
    return state;
  }

  if (action.payload.field === 'on') {
    return state.map(filter =>
      (filter.text === action.payload.text) ? { ...filter, on: !filter.on } : filter);
  }

  if (action.payload.field === 'inverse') {
    return state.map(filter =>
      (filter.text === action.payload.text) ? { ...filter, inverse: !filter.inverse } : filter);
  }

  if (action.payload.field === 'remove') {
    return state.filter(filter =>
      (filter.text !== action.payload.text));
  }

  if (action.payload.field === 'add') {
    return [
      ...state,
      {
        text: action.payload.text,
        on: true,
        inverse: false
      }
    ];
  }

  return state;
}
