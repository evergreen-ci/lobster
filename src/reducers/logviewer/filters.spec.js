import * as actions from '../../actions';
import filters from './filters';

describe('filters', () => {
  test('add', () => {
    const out = filters([], actions.addFilter('text', true));
    expect(out).toHaveLength(1);
    expect(out[0]).toMatchSnapshot();

    const out2 = filters(out, actions.addFilter('text', false));
    expect(out2).toHaveLength(1);
    expect(out2[0].caseSensitive).toBe(true);
  });

  test('caseSensitive', () => {
    const out = filters(filters([], actions.addFilter('text', true)), actions.toggleFilterCaseSensitive('text'));
    expect(out).toHaveLength(1);
    expect(out[0].caseSensitive).toBe(false);

    const out2 = filters(out, actions.toggleFilterCaseSensitive('text'));
    expect(out2).toHaveLength(1);
    expect(out2[0].caseSensitive).toBe(true);
  });

  test('remove', () => {
    const out = filters([], actions.addFilter('text', true));
    expect(out).toHaveLength(1);

    const out2 = filters(out, actions.removeFilter('text'));
    expect(out2).toHaveLength(0);
  });

  test('inverse', () => {
    const out = filters([], actions.addFilter('text', true));
    expect(out[0].inverse).toBe(false);

    const out2 = filters(out, actions.toggleFilterInverse('text'));
    expect(out2[0].inverse).toBe(true);

    const out3 = filters(out2, actions.toggleFilterInverse('text'));
    expect(out3[0].inverse).toBe(false);
  });

  test('on', () => {
    const out = filters([], actions.addFilter('text', true));
    expect(out[0].on).toBe(true);

    const out2 = filters(out, actions.toggleFilter('text'));
    expect(out2[0].on).toBe(false);

    const out3 = filters(out2, actions.toggleFilter('text'));
    expect(out3[0].on).toBe(true);
  });

  test('load', () => {
    const out = filters([], actions.loadInitialFilters([
      {
        'text': 'doop',
        'on': true,
        'line': 'true',
        'caseSensitive': true
      },
      {
        'text': 'boop',
        'on': true,
        'line': 'true',
        'caseSensitive': true
      }
    ]));
    expect(out).toHaveLength(2);
  });
});
