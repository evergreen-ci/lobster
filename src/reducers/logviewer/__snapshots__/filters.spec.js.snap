// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters add 1`] = `
Object {
  "caseSensitive": true,
  "inverse": false,
  "on": true,
  "text": "text",
}
`;
